File: src/release-notes/OutputWriter.fs
namespace ReleaseNotes

open System
open System.IO
open System.Text

type OutputWriter(output: string option) =
    let stdout = Console.Out
    let sb = new StringBuilder()
    do
        output |> Option.iter (fun o ->
            printfn ""
            printfn "Building %s" o
            printfn "-------------------"   
        )
    
    member this.EmptyLine () =
        stdout.WriteLine()
        sb.AppendLine() |> ignore
    member this.WriteLine (s:string) =
        stdout.WriteLine(s)
        sb.AppendLine(s) |> ignore
    override this.ToString () = sb.ToString()
    interface IDisposable with 
        member __.Dispose() =
            stdout.Dispose()
            match output with
            | None -> ()
            | Some path ->
                let fi = FileInfo(path)
                Directory.CreateDirectory(fi.Directory.FullName) |> ignore
                File.WriteAllText(path, sb.ToString())
        

File: src/release-notes/OutputWriter.cs
using System.Text;

namespace ReleaseNotes;

public sealed class OutputWriter : IDisposable
{
	private readonly TextWriter _stdout = Console.Out;
	private readonly StringBuilder _sb = new();
	private readonly string? _output;

	public OutputWriter(string? output)
	{
		_output = output;
		if (output is null) return;

		Console.WriteLine();
		Console.WriteLine($"Building {output}");
		Console.WriteLine("-------------------");
	}

	public void EmptyLine()
	{
		_stdout.WriteLine();
		_sb.AppendLine();
	}

	public void WriteLine(string s)
	{
		_stdout.WriteLine(s);
		_sb.AppendLine(s);
	}

	public override string ToString() => _sb.ToString();

	public void Dispose()
	{
		if (_output is not null)
		{
			var fi = new FileInfo(_output);
			Directory.CreateDirectory(fi.DirectoryName!);
			File.WriteAllText(_output, _sb.ToString());
		}
	}
}
